File: solution/0300-0399/0323.Number of Connected Components in an Undirected Graph/Solution.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        p = [i for i in range(n)]

        def find(x):
            if p[x] != x:
                p[x] = find(p[x])
            return p[x]

        for a, b in edges:
            p[find(b)] = find(a)
        return sum(i == find(i) for i in range(n))

File: solution/0300-0399/0323.Number of Connected Components in an Undirected Graph/Solution.java
class Solution {
    private int[] p;
    
    public int countComponents(int n, int[][] edges) {
        p = new int[n];
        for (int i = 0; i < n; ++i) {
            p[i] = i;
        }
        for (int[] e : edges) {
            int a = e[0], b = e[1];
            p[find(b)] = find(a);
        }

        int cnt = 0;
        for (int i = 0; i < n; ++i) {
            if (i == find(i)) {
                ++cnt;
            }
        }
        return cnt;
    }

    private int find(int x) {
        if (p[x] != x) {
            p[x] = find(p[x]);
        }
        return p[x];
    }
}